### ISSUE TITLE
Function naming problems

### ISSUE THREAD

**Issue by stefanmoosbrugger** [insieme/insieme#61]
There are several problems in the naming of member functions.

**1) Pure virtual member functions may be handled incorrectly.** To give you one example 
have a look on the following code.

```
class A {
public:
  virtual void test() = 0;
};
class A : public B {
public:
  void test() { ... }
};
int main() {
  B obj;
  obj.test();
  return 0;
}
```
In this case the produced backend code contains the ```test``` method only in class B. 
The pure virtual information is lost. This is not a big deal, because the backend code is
compilable, but still though I think it should be done right. 

Anyway.. Lets try to do a cast to the base class (ptr). This will lead to correct output code. So 
everything is there but this leads to the next problem...

**2) Combination of Const method and pure virtual does not work (or more general: we should get rid of our bad "name keeping conventions").** Again have a look at the following example:
```
class A {
public:
  virtual void test() const = 0;
};
class A : public B {
public:
  void test() const { ... }
};
int main() {
  B obj;
  A* a_obj = &obj;
  a_obj->test();
  ...
  return 0;
}
```
The problem is that const methods are extended with ```_c```. This dirty fix 
was done (afaik) one or two years ago, because of incorrect code generation 
when writing functions that only differ by const. The output code in this case will look
like:
```
struct A {
  virtual void test() const = 0;
};
struct B {
  void test_c() const { ... }
};

...
```
Does anyone have a nice (or better) solution for this problem?! I'm afraid that this problem does not only occur when using pure virtual and const methods. The real problem is that we want to keep the name for some code elements (e.g., (pure) virtual methods). This is a very dangerous practice Unfortunately I don't remember further examples but I know that there have been several problems with this "name keeping convention" so far. 

**2.1) Further example**
Just to provide you a more easy example.. Imagine you have a class 
that contains following member function prototypes:
```
void fun() const;
void fun();
```
At the moment we use the renaming that is explained in 2 which will
generate backend code that looks somehow like this:
```
void fun_c() const;
void fun();
```

**Comment by HerbertJordan** [comment]
Do we have some documentation of the handling of C++ classes and member functions accurately enough so that we can use it as a basis for cleaning up their encoding?

**Comment by stefanmoosbrugger** [comment]
In my opinion: not really. The messy naming infrastructure more or 
less evolved because several (c++) language elements got implemented 
(e.g., pure virtual) and every now and then we had to fix bugs in the
name generating utility.

The only good thing is that "most" naming functionalities are in one single
name utility (see https://github.com/insieme/insieme/blob/master/code/frontend/src/utils/name_manager.cpp). This utility is used to generate names for enums, (templated) functions, etc.

I guess the easiest and most efficient way to fix (hopefully) all of our naming issued would be
to drop the name manager and write it from scratch. Unfortunately I just discovered another bug that is related with enum naming, therefore I will mark this as a severe bug that should be fixed asap.

**Comment by PeterTh** [comment]
Looking at the name manager, I find it a bit concerning that operators are only recognized by their names. Isn't this semantic information which shouldn't be hidden in what is ultimately an annotation?

**Comment by HerbertJordan** [comment]
I was re-reading through this issue ... kind of a cold-case thing .. and thinking that we could distinguish const-functions by passing a src as a this reference, while having non-member functions having a reference as a first argument. This would integrate the const-modifier into the type of the function and we can get rid of this ugly name-mods.

Do you agree?

**Comment by HerbertJordan** [comment]
sorry

**Comment by PeterTh** [comment]
As far as I can tell, the root causes of all of these issues were resolved in the inspire 2.0 redesign.
